rust/bridge/shared/src/node/convert.rs

```diff
@@ -632,47 +631,170 @@ pub(crate) unsafe fn extend_lifetime<'a, 'b: 'a, T: ?Sized>(some_ref: &'a T) ->
 /// The name of the property on JavaScript objects that wrap a boxed Rust value.
 pub(crate) const NATIVE_HANDLE_PROPERTY: &str = "_nativeHandle";
 
+/// A marker trait for Rust objects exposed to JavaScript through [`neon::types::JsBox`].
+///
+/// Since this trait is used as a marker and implemented through a macro,
+/// operations that might differ across types have been factored into the [`BridgeHandleStrategy`]
+/// trait. The [`bridge_handle`] macro will automatically choose the correct strategy.
+pub trait BridgeHandle: Send + Sync + Sized + 'static {
+    /// Factors out operations that differ between mutable and immutable bridge handles.
+    type Strategy: BridgeHandleStrategy<Self>;
+}
+
+/// Factors out operations that differ between mutable and immutable bridge handles.
+///
+/// See the [`BridgeHandle`] trait.
+pub trait BridgeHandleStrategy<T> {
+    /// If the Rust object needs any additional wrapping before being boxed by Neon,
+    /// this type can be used.
+    type JsBoxContents: From<T> + Send + 'static;
+}
+
+/// A phantom type for immutable bridge handles.
+///
+/// See [`BridgeHandleStrategy<T>`].
+pub struct Immutable<T>(std::marker::PhantomData<T>);
+
+/// A phantom type for mutable bridge handles.
+///
+/// See [`BridgeHandleStrategy<T>`].
+pub struct Mutable<T>(std::marker::PhantomData<T>);
+
+/// Immutable handles need no extra boxing.
+impl<T: BridgeHandle> BridgeHandleStrategy<T> for Immutable<T> {
+    type JsBoxContents = T;
+}
+
+/// Mutable handles use a [`RefCell`] for safety.
+impl<T: BridgeHandle> BridgeHandleStrategy<T> for Mutable<T> {
+    type JsBoxContents = RefCell<T>;
+}
+
+/// Shorthand for `T::Strategy::JsBoxContents` when `T: `[`BridgeHandle`], which Rust can't always
+/// resolve.
+type BoxContentsFor<T> = <<T as BridgeHandle>::Strategy as BridgeHandleStrategy<T>>::JsBoxContents;
+
+impl<'a, T: BridgeHandle> ResultTypeInfo<'a> for T {
+    type ResultType = JsValue;
+    fn convert_into(self, cx: &mut impl Context<'a>) -> NeonResult<Handle<'a, Self::ResultType>> {
+        Ok(cx
+            .boxed(DefaultFinalize(BoxContentsFor::<T>::from(self)))
+            .upcast())
+    }
+}
+/// Accesses a boxed Rust value using a custom reference (such as [`std::cell::RefMut`]).
+///
+/// Once created, the value can be accessed using the usual [`Deref`] and [`DerefMut`] operations.
+///
+/// Also stores the JavaScript handle that owns the JsBox to ensure that the Rust value is kept
+/// alive.
+pub struct BoxedBorrow<'a, Borrow: Deref + 'a>
```

Comment:
I know it's explicit, but I think this would be much more clear with a name like BorrowedJsBoxedBridgeHandle.  It took me a long time reading this code to figure out what was going on, and I think that name would have helped a lot. In particular, to make it clear that we're borrowing a BridgeHandle that's in a JsBox.

Reply:
Similarly, I know it's explicit, but instead of "Borrow" being the type name, having it be  BorrowedBoxHandle would have made this a lot more clear.

Reply:
I think the comment for the class would be more clear with something like

"Contains a JsBoxed BridgeHandle along with a Borrowed version of it.  Can technically borrow from anything of Deref<Target=BridgeHandle> but in practice borrows from a &BridgeHandle,  a RefCell<BridgeHandle>::Ref,  or a RefCell<BridgeHandle>::MutRef"